### PR TITLE
Some changes for OML 3.0

### PR DESCRIPTION
**Pull request by AlekseySh**
**CHANGELOG**

* Moved category wise metrics calculation logic from `EmbeddingMetrics` to functional metrics.
* Removed `fnmr@fmr` metric from `EmbeddingMetrics` because we cannot guarantee correctness of its behaviour when postprocessor is presented and the metric is computationally heavy. [decided not to remove this metric]
* Reworked handling empty bboxes (use one None instead of 4 Nones)
* `calc_retrieval_metrics_on_full`, `calc_gt_mask`, `calc_mask_to_ignore`, `apply_mask_to_ignore` finally moved to tests to serve as adapters between the old and the new ways of computing metrics
* pipelines: a bit of refactoring and improved type hints
* added `show` argument to `RetrievalResults.visualise()`

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -345,16 +347,19 @@ from oml.registry.transforms import get_transforms_for_pretrained
 extractor = ViTExtractor.from_pretrained("vits16_dino")
 transform, _ = get_transforms_for_pretrained("vits16_dino")
 
-_, df_val = download_mock_dataset(global_paths=True)
+_, df_val = download_mock_dataset(global_paths=True, df_name="df_with_category.csv")
 dataset = ImageQueryGalleryLabeledDataset(df_val, transform=transform)
 
+# you can optionally provide categories to have category wise metrics
+query_categories = np.array(dataset.extra_data["category"])[dataset.get_query_ids()]
```
Comment (AlekseySh) [inline README.md:354]: does it seem okay in the examples?

**Comment by deepslug** [comment]
@AlekseySh 

> Removed fnmr@fmr metric from EmbeddingMetrics because we cannot guarantee correctness of its behaviour when postprocessor is presented and the metric is computationally heavy.

Is it possible to keep this metric available as an option if user wants to use it, because this metric is considered the most important metric in the field of biometrics (e.g., face/fingerprint recoginition) where score thresholding is often employed?

Having said that, it is also super helpful if this metric (or any other "the-lower-the-better" metrics) can be specified to metric_for_checkpointing like "OVERALL/fnmr@fmr/0.001" with the "mode" as "min" via YAML. The mode is hard-coded as "max" in the current implementation:

```
def parse_ckpt_callback_from_config(cfg: TCfg) -> ModelCheckpoint:
    return ModelCheckpoint(
        dirpath=Path.cwd() / "checkpoints",
        monitor=cfg["metric_for_checkpointing"],
        mode="max",
        save_top_k=1,
        verbose=True,
        filename="best",
    )
```

File: oml/datasets/images.py [hunk 2]
```diff
@@ -162,8 +161,9 @@ def __len__(self) -> int:
         return len(self._paths)
 
     def visualize(self, item: int, color: TColor = BLACK) -> np.ndarray:
-        bbox = torch.tensor(self._bboxes[item]) if (self._bboxes is not None) else torch.tensor([torch.nan] * 4)
-        image = get_img_with_bbox(im_path=self._paths[item], bbox=bbox, color=color)
+        img = np.array(imread_pillow(self.read_bytes(self._paths[item])))
```
Comment (DaloroAT) [inline oml/datasets/images.py:164]: you can read with `cv2` directly to `np`. In addition it's faster

File: oml/functional/metrics.py [hunk 3]
```diff
@@ -57,51 +61,30 @@ def calc_retrieval_metrics(
         metrics["precision"] = dict(zip(precision_top_k, precision))
 
     if map_top_k:
-        map = calc_map(gt_tops, n_gts, map_top_k)
-        metrics["map"] = dict(zip(map_top_k, map))
+        map_ = calc_map(gt_tops, n_gts, map_top_k)
```
Comment (DaloroAT) [inline oml/functional/metrics.py:64]: respect built-in

File: oml/functional/metrics.py [hunk 4]
```diff
@@ -57,51 +61,30 @@ def calc_retrieval_metrics(
...
+        metrics["map"] = dict(zip(map_top_k, map_))
+
+    if query_categories is not None:
+        metrics_cat = {c: take_unreduced_metrics_by_mask(metrics, query_categories == c) for c in query_categories}
```
Comment (DaloroAT) [inline oml/functional/metrics.py:68]: A bit strange to have a simple dict `metrics` but metrics keys are mixed with some categories keys on the same top level

```python
metrics["map"][5] = 0.5
metrics["cmc"][3] = 0.4
metrics["OVERALL"]["cmc"][3] = 0.4
metrics["cats"]["cmc"][3] = 0.1
metrics["pigs"]["cmc"][3] = 0.2
...
```

File: oml/metrics/embeddings.py [hunk 5]
```diff
@@ -143,19 +135,17 @@ def update_data(self, data: Dict[str, Any], indices: TIndices) -> Any:
...
 
         if self.postprocessor:
             self.retrieval_results = self.postprocessor.process(self.retrieval_results, self.dataset)
 
-    def compute_metrics(self) -> TMetricsDict_ByLabels:  # type: ignore
-        self.acc = self.acc.sync()  # if DDP gathering happens here
-
-        # todo 522: put back fnmr metric
+    def compute_metrics(self) -> TMetricsDict:  # type: ignore
+        self.acc = self.acc.sync()  # gathering data from devices happens here if DDP
```
Comment (DaloroAT) [inline oml/metrics/embeddings.py:148]: wrong comment

File: oml/metrics/embeddings.py [hunk 6]
```diff
@@ -166,47 +156,42 @@ def compute_metrics(self) -> TMetricsDict_ByLabels:  # type: ignore
...
 
-            for category in np.unique(query_categories):
-                mask_query_sz = query_categories == category
-                metrics[category] = take_unreduced_metrics_by_mask(metrics[self.overall_categories_key], mask_query_sz)
+            metrics_r = calc_retrieval_metrics_rr(query_categories=query_categories, **args_r)  # type: ignore
+            metrics_t = calc_topological_metrics(categories=categories, **args_t)  # type: ignore
 
-                mask_dataset_sz = categories == category
-                metrics[category].update(calc_topological_metrics(embeddings[mask_dataset_sz], self.pcf_variance))
+            self.metrics_unreduced = {cat: {**metrics_r[cat], **metrics_t[cat]} for cat in metrics_r.keys()}
```
Comment (DaloroAT) [inline oml/metrics/embeddings.py:176]: On this line `metrics_r` have all retrieval metrics and categories as top-level keys because of [this](https://github.com/OML-Team/open-metric-learning/blob/521df23cb0e9231a48f0251da95106a413e5afac/oml/functional/metrics.py#L67), so
1) You can't do `metrics_t[cat]` for retrieval metric keys
2) `self.metrics_unreduced` won't include `metrics_t`

**Comment by AlekseySh** [comment]
@deepslug okay, I got your point

File: tests/test_oml/test_functional/test_metrics/test_retrieval_metrics.py [hunk 7]
```diff
@@ -275,3 +280,30 @@ def test_calc_fnmr_at_fmr_check_params(fmr_vals: Tuple[int, ...]) -> None:
...
+    metrics = calc_retrieval_metrics(
+        retrieved_ids=retrieved_ids,
+        gt_ids=gt_ids,
+        query_categories=query_categories,
+        cmc_top_k=(1,),
+        precision_top_k=(3, 5),
+        map_top_k=tuple(),
+    )
+
+    assert math.isclose(metrics["cat"]["cmc"][1], 1)
```
Comment (AlekseySh) [inline tests/test_oml/test_functional/test_metrics/test_retrieval_metrics.py:299]: rework this assers: compare dicts instead (to check if there are unwilling extra keys)

File: oml/functional/metrics.py (hunk 4 above)
Comment (AlekseySh) [reply oml/functional/metrics.py:68]: no, it's:

```
{
        "cat": {"cmc": {1: 1.0}, "precision": {3: 2 / 3, 5: 2 / 3}},
        "dog": {"cmc": {1: 1.0}, "precision": {3: 1 / 2, 5: 1 / 2}},
        OVERALL_CATEGORIES_KEY: ...,
    }
```

File: oml/datasets/images.py (hunk 2 above)
Comment (AlekseySh) [reply oml/datasets/images.py:164]: I did it to support paths as urls later on

File: oml/metrics/embeddings.py (hunk 6 above)
Comment (AlekseySh) [reply oml/metrics/embeddings.py:176]: discussed offline

**Comment by AlekseySh** [comment]
Changes from this PR have been moved to other PRs:
* https://github.com/OML-Team/open-metric-learning/pull/568
* https://github.com/OML-Team/open-metric-learning/pull/567
* https://github.com/OML-Team/open-metric-learning/pull/566
